### PR TITLE
Feature/add authentication api

### PR DESCRIPTION
**Pull request by wave49192**
Implemented
- Authentication api with sign in, sign out, register
- Protect other route by checking the cookie token first

### CODE REVIEW

File: app/api/v1/auth_api.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    resource :auth do
+      desc "Sign in a user"
+      params do
+        requires :email, type: String, desc: "User's email"
+        requires :password, type: String, desc: "User's password"
+      end
+      post "sign_in" do
+        user = authenticate_user!(params[:email], params[:password])
+        cookies[:access_token] = {
```
Comment (siripoom-datawow) [inline app/api/v1/auth_api.rb:37]: plesae seperate as a method since it can be reused

File: app/api/v1/auth_api.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        user = User.new(email: email, password: password, password_confirmation: password_confirmation)
+        if user.save
+          user.regenerate_access_token
+          user.save
+          user
+        else
+          error!(user.errors.full_messages, 422)
+        end
+      end
+    end
```
Comment (siripoom-datawow) [inline app/api/v1/auth_api.rb:27]:
```suggestion
def create_user!(email, password, password_confirmation)
  user = User.new(email: email, password: password, password_confirmation: password_confirmation)
  return error!(user.errors.full_messages, 422) unless user.save
  
  user.regenerate_access_token
  user.save
  user
end

    end
```

File: app/api/v1/auth_api.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          user.regenerate_access_token
+          user.save
+          user
+        else
+          error!("Invalid Email or Password", 401)
+        end
+      end
+
+      def create_user!(email, password, password_confirmation)
+        user = User.new(email: email, password: password, password_confirmation: password_confirmation)
```
Comment (siripoom-datawow) [inline app/api/v1/auth_api.rb:18]: should it validate existed user?

File: app/api/v1/auth_api.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        if token.nil?
+          error!("Unauthorized, no access token provided", 401)
+        end
+        current_user = User.find_by(access_token: token)
+        if current_user
+          current_user.update(access_token: nil)
+          cookies.delete(:access_token)
+          { message: "Logged out successfully" }
+        else
+          error!("Unauthorized, invalid token", 401)
```
Comment (siripoom-datawow) [inline app/api/v1/auth_api.rb:57]: also can use guarding here

File: app/api/v1/auth_api.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
+module V1
+  class AuthApi < Grape::API
+    format :json
+
+    helpers do
+      def authenticate_user!(email, password)
+        user = User.find_by(email: email)
+        if user && user.valid_password?(password)
+          user.regenerate_access_token
+          user.save
```
Comment (NetStriKerX) [inline app/api/v1/auth_api.rb:10]: I think `user.regenerate_access_token` will updated the record, so no need to add `user.save` here.

We do `record.save` when we change value of attribute this way ==> `user.email = 'new_email@gmail.com'` (won't update record until `user.save`).

File: app/api/v1/auth_api.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+      desc "Register a new user"
+      params do
+        requires :email, type: String, desc: "User's email"
+        requires :password, type: String, desc: "User's password"
+        requires :password_confirmation, type: String, desc: "User's password confirmation"
+      end
+
+      post "register" do
+        user = create_user!(params[:email], params[:password], params[:password_confirmation])
```
Comment (kaiitoon) [inline app/api/v1/auth_api.rb:69]: Maybe check password and password confirmation is matched

File: app/api/v1/auth_api.rb (hunk 3 above)
Comment (wave49192) [reply app/api/v1/auth_api.rb:18]: Devise does that job for us
in `user.rb`
```  
devise :database_authenticatable, :registerable,
        
       :recoverable, :rememberable, :validatable
```
the `:validatable` module checks the email and password validations

File: app/api/v1/auth_api.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
+module V1
+  class AuthApi < Grape::API
+    format :json
+
+    helpers do
```
[the remark this answers is not in the document]
Comment (wave49192) [reply app/api/v1/auth_api.rb:5]: error occurs when move helper to the bottom off the page 
`undefined method 'helpers' for module V1`

it works fine only when defined before the `resource :auth do` line

so I think it should defined before the actual endpoint logic

File: app/api/v1/auth_api.rb (hunk 6 above)
Comment (wave49192) [reply app/api/v1/auth_api.rb:69]: Devise does that job for us
in `user.rb`
```  
devise :database_authenticatable, :registerable,
        
       :recoverable, :rememberable, :validatable
```
the `:validatable` module checks the email and password validations

**Pull request merged**